## **Implementation of Dijkstra and A star algorithm**

In [18]:
import heapq
import math

class solution:
    def __init__(self):
        self.grid = [[0 for _ in range(201)] for _ in range(201)]

    def addObs(self, lx, ly, rx, ry):
        for i in range(lx, rx + 1):
            for j in range(ly, ry + 1):
                if i >= 0 and i < 201 and j >= 0 and j < 201:
                    self.grid[i][j] = 1

    def checkval(self, x, y):
        if x >= 0 and x < 201 and y >= 0 and y < 201 and self.grid[x][y] != 1:
            return True
        return False

    def dijkstra(self):
        start = (0, 0)
        end = (200, 200)
        distances = [[float('inf') for _ in range(201)] for _ in range(201)]
        visited = [[0 for _ in range(201)] for _ in range(201)]
        parent = [[None for _ in range(201)] for _ in range(201)]
        distances[0][0] = 0
        pq = [(0, start)]
        cells_explored = 0
        while pq:
            current_dist,(x,y) = heapq.heappop(pq)
            if visited[x][y] == 1:
                continue
            visited[x][y] = 1
            cells_explored += 1
            if x == end[0] and y == end[1]:
                break
            neighbors = [(1, 0, 1), (-1, 0, 1), (0, 1, 1), (0, -1, 1), (1, 1, math.sqrt(2)), (1, -1, math.sqrt(2)), (-1, 1, math.sqrt(2)), (-1, -1, math.sqrt(2))]
            for dx, dy, weight in neighbors:
                nx, ny = x + dx, y + dy
                if self.checkval(nx, ny):
                    new_dist = current_dist + weight
                    if new_dist < distances[nx][ny]:
                        distances[nx][ny] = new_dist
                        parent[nx][ny] = (x, y)
                        heapq.heappush(pq, (new_dist, (nx, ny)))

        path = []
        curr = end
        while curr is not None:
            path.append(curr)
            curr = parent[curr[0]][curr[1]]
        path.reverse()

        print(f"Number of cells explored: {cells_explored}")
        print(f"Shortest distance: {distances[end[0]][end[1]]}")
        print("Path:")
        print(path)
        print("Length of path:",len(path))

    def heuristic(self, current, goal):
        dx = abs(current[0] - goal[0])
        dy = abs(current[1] - goal[1])
        return (dx + dy) + (math.sqrt(2) - 2) * min(dx, dy)

    def a_star(self):
        start = (0, 0)
        end = (200, 200)
        g_score = [[float('inf') for _ in range(201)] for _ in range(201)]
        g_score[0][0] = 0
        pq = [(self.heuristic(start, end), start)]
        visited = [[False for _ in range(201)] for _ in range(201)]
        parent = [[None for _ in range(201)] for _ in range(201)]
        cells_explored = 0
        while pq:
            current_f,(x,y) = heapq.heappop(pq)
            if visited[x][y]:
                continue
            visited[x][y] = True
            cells_explored+=1
            if (x,y) == end:
                break
            neighbors = [(1, 0, 1),(-1, 0, 1),(0, 1, 1),(0, -1, 1),(1, 1, math.sqrt(2)),(1, -1, math.sqrt(2)),(-1, 1, math.sqrt(2)),(-1, -1, math.sqrt(2))]
            for dx,dy,weight in neighbors:
                nx,ny = x+dx,y+dy
                if self.checkval(nx, ny):
                    tentative_g = g_score[x][y] + weight
                    if tentative_g < g_score[nx][ny]:
                        g_score[nx][ny] = tentative_g
                        parent[nx][ny] = (x, y)
                        f_score = tentative_g + self.heuristic((nx,ny),end)
                        heapq.heappush(pq,(f_score,(nx,ny)))
        path = []
        curr = end
        while curr is not None:
            path.append(curr)
            curr = parent[curr[0]][curr[1]]
        path.reverse()
        print(f"Number of cells explored: {cells_explored}")
        print(f"Shortest distance: {g_score[end[0]][end[1]]}")
        print("Path:")
        print(path)
        print("Length of path:",len(path))


In [19]:
s = solution()
s.addObs(40, 40, 70, 70)
s.addObs(90, 120, 130, 160)
s.addObs(150, 50, 180, 90)
s.addObs(60, 150, 100, 190)

# **Results**

In [20]:
s.dijkstra()

Number of cells explored: 34928
Shortest distance: 301.002092041054
Path:
[(0, 0), (1, 0), (2, 0), (3, 0), (4, 0), (5, 0), (6, 0), (7, 0), (8, 0), (9, 0), (10, 0), (11, 0), (12, 0), (13, 0), (14, 0), (15, 0), (16, 0), (17, 0), (18, 0), (19, 0), (20, 0), (21, 0), (22, 0), (23, 0), (24, 0), (25, 0), (26, 0), (27, 0), (28, 0), (29, 0), (30, 0), (31, 0), (32, 1), (33, 2), (34, 3), (35, 4), (36, 5), (37, 6), (38, 7), (39, 8), (40, 9), (41, 10), (42, 11), (43, 12), (44, 13), (45, 14), (46, 15), (47, 16), (48, 17), (49, 18), (50, 19), (51, 20), (52, 21), (53, 22), (54, 23), (55, 24), (56, 25), (57, 26), (58, 27), (59, 28), (60, 29), (61, 30), (62, 31), (63, 32), (64, 33), (65, 34), (66, 35), (67, 36), (68, 37), (69, 38), (70, 39), (71, 40), (71, 41), (71, 42), (72, 43), (73, 44), (74, 45), (75, 46), (76, 47), (77, 48), (78, 49), (79, 50), (80, 51), (81, 52), (82, 53), (83, 54), (84, 55), (85, 56), (86, 57), (87, 58), (88, 59), (89, 60), (90, 61), (91, 62), (92, 63), (93, 64), (94, 65), (95, 6

In [21]:
s.a_star()

Number of cells explored: 8134
Shortest distance: 301.002092041054
Path:
[(0, 0), (1, 1), (2, 2), (3, 3), (4, 4), (5, 5), (6, 6), (7, 6), (8, 6), (9, 6), (10, 6), (11, 6), (12, 6), (13, 6), (14, 6), (15, 6), (16, 6), (17, 6), (18, 6), (19, 6), (20, 6), (21, 6), (22, 6), (23, 6), (24, 6), (25, 6), (26, 6), (27, 6), (28, 6), (29, 6), (30, 6), (31, 7), (32, 7), (33, 8), (34, 9), (35, 10), (36, 11), (37, 12), (38, 13), (39, 14), (40, 15), (41, 15), (42, 16), (43, 17), (44, 18), (45, 19), (46, 19), (47, 20), (48, 21), (49, 22), (50, 23), (51, 23), (52, 24), (53, 25), (54, 26), (55, 27), (56, 27), (57, 28), (58, 29), (59, 30), (60, 31), (61, 31), (62, 32), (63, 33), (64, 34), (65, 35), (66, 35), (67, 36), (68, 37), (69, 38), (70, 39), (71, 40), (72, 41), (73, 42), (74, 43), (75, 44), (76, 45), (77, 46), (78, 47), (79, 48), (80, 49), (81, 50), (82, 51), (83, 52), (84, 53), (85, 54), (86, 55), (87, 56), (88, 57), (89, 58), (90, 59), (91, 60), (92, 61), (93, 62), (94, 63), (95, 64), (96, 65), (

## **Analysis**

**1.Heuristic admissibility:** The Octile distance heuristic is admissible because it calculates the exact shortest distance between two points in an 8-way grid assuming no obstacles. Since obstacles can only increase the actual path length, the heuristic never overestimates the true remaining cost, ensuring an optimal solution.

**2.Path length:** The path length remains identical at approximately 301.002092041054 units. This is because both Dijkstra and A* are designed to be optimal solvers. Since the heuristic used for A* was admissible (meaning it never overestimates the distance), it guaranteed the same shortest path distance as Dijkstra.

**3.Number of expanded nodes** Dijkstra expanded 34,928 nodes, while A* only expanded 8,134. A* reduced the search space by roughly 77% because it uses a heuristic to prioritize nodes closer to the goal, whereas Dijkstra searches in every direction.

**4.Overall efficiency:** A* is the much faster and smarter choice here. It gets the same result as Dijkstra but uses a fraction of the memory and processing power. By focusing on the target instead of searching the entire map, it reaches the destination with much less effort.